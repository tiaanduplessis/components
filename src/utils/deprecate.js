/* eslint-disable no-console */
import {useRef, useCallback} from 'react'

const noop = () => {}
// eslint-disable-next-line import/no-mutable-exports
let deprecate = null

if (__DEV__) {
  deprecate = ({name, message, version}) => {
    Deprecations.deprecate({name, message, version})
  }
} else {
  deprecate = noop
}

export {deprecate}

// eslint-disable-next-line import/no-mutable-exports
let useDeprecation = null

if (__DEV__) {
  useDeprecation = ({name, message, version}) => {
    const ref = useRef(false)
    const logDeprecation = useCallback(() => {
      if (!ref.current) {
        ref.current = true
        deprecate({name, message, version})
      }
    }, [name, message, version])

    return logDeprecation
  }
} else {
  useDeprecation = () => {
    return noop
  }
}

export {useDeprecation}

export class Deprecations {
  static get() {
    if (!Deprecations.instance) {
      Deprecations.instance = new Deprecations()
    }

    return Deprecations.instance
  }

  constructor() {
    this.deprecations = []
  }

  static deprecate({name, message, version}) {
    const msg = `WARNING! ${name} is deprecated and will be removed in version ${version}. ${message}`
    console.warn(msg)

    this.get().deprecations.push({name, message, version})
  }

  static getDeprecations() {
    return this.get().deprecations
  }

  static clearDeprecations() {
    this.get().deprecations.length = 0
  }
}
