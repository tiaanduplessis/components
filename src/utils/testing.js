/* eslint-disable jsdoc/require-param, jsdoc/require-returns, jsdoc/require-description-complete-sentence, jsdoc/match-description */
import React from 'react'
import {promisify} from 'util'
import renderer from 'react-test-renderer'
import enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import {ThemeProvider} from 'styled-components'
import {default as defaultTheme} from '../theme'

const readFile = promisify(require('fs').readFile)

export const COMPONENT_DISPLAY_NAME_REGEX = /^[A-Z][A-Za-z]+(\.[A-Z][A-Za-z]+)*$/

enzyme.configure({adapter: new Adapter()})

export function mount(component) {
  return enzyme.mount(component)
}

/**
 * Render the component (a React.createElement() or JSX expression)
 * into its intermediate object representation with 'type',
 * 'props', and 'children' keys
 *
 * The returned object can be matched with expect().toEqual(), e.g.
 *
 * ```js
 * expect(render(<Foo />)).toEqual(render(<div foo='bar' />))
 * ```
 */
export function render(component) {
  return renderer.create(component).toJSON()
}

/**
 * Render the component (a React.createElement() or JSX expression)
 * using react-test-renderer and return the root node
 * ```
 */
export function renderRoot(component) {
  return renderer.create(component).root
}

/**
 * Get the HTML class names rendered by the component instance
 * as an array.
 *
 * ```js
 * expect(renderClasses(<div className='a b' />))
 *   .toEqual(['a', 'b'])
 * ```
 */
export function renderClasses(component) {
  const {
    props: {className}
  } = render(component)
  return className ? className.trim().split(' ') : []
}

/**
 * Returns true if a node renders with a single class.
 */
export function rendersClass(node, klass) {
  return renderClasses(node).includes(klass)
}

export function renderWithTheme(node, theme = defaultTheme) {
  return render(<ThemeProvider theme={theme}>{node}</ThemeProvider>)
}

export function px(value) {
  return typeof value === 'number' ? `${value}px` : value
}

export function percent(value) {
  return typeof value === 'number' ? `${value}%` : value
}

export function renderStyles(node) {
  const {
    props: {className}
  } = render(node)
  return getComputedStyles(className)
}

export function getComputedStyles(className) {
  const div = document.createElement('div')
  div.className = className

  const computed = {}
  for (const sheet of document.styleSheets) {
    for (const rule of sheet.cssRules) {
      if (rule.type === 1) {
        readRule(rule, computed)
      } else if (rule.type === 4) {
        readMedia(rule)
      } else {
        // console.warn('rule.type =', rule.type)
      }
    }
  }

  return computed

  function readRule(rule, dest) {
    if (matchesSafe(div, rule.selectorText)) {
      const {style} = rule
      for (let i = 0; i < style.length; i++) {
        const prop = style[i]
        dest[prop] = style.getPropertyValue(prop)
      }
    } else {
      // console.warn('no match:', rule.selectorText)
    }
  }

  function readMedia(mediaRule) {
    const key = `@media ${mediaRule.media[0]}`
    // const dest = computed[key] || (computed[key] = {})
    const dest = {}
    for (const rule of mediaRule.cssRules) {
      readRule(rule, dest)
    }

    // Don't add media rule to computed styles
    // if no styles were actually applied
    if (Object.keys(dest).length > 0) {
      computed[key] = dest
    }
  }

  function matchesSafe(node, selector) {
    if (!selector) {
      return false
    }
    try {
      return node.matches(selector)
    } catch (error) {
      return false
    }
  }
}

/**
 * This provides a layer of compatibility between the render() function from
 * react-test-renderer and Enzyme's mount()
 */
export function getProps(node) {
  return typeof node.props === 'function' ? node.props() : node.props
}

export function getClassName(node) {
  return getProps(node).className
}

export function getClasses(node) {
  const className = getClassName(node)
  return className ? className.trim().split(/ +/) : []
}

export function loadCSS(path) {
  return readFile(require.resolve(path), 'utf8').then(css => {
    const style = document.createElement('style')
    style.setAttribute('data-path', path)
    style.textContent = css
    document.head.appendChild(style)
    return style
  })
}

export function unloadCSS(path) {
  const style = document.querySelector(`style[data-path="${path}"]`)
  if (style) {
    style.remove()
    return true
  }
}

// If a component requires certain props or other conditions in order
// to render without errors, you can pass a `toRender` function that
// returns an element ready to be rendered.
export function behavesAsComponent(Component, systemPropArray, toRender = null, options) {
  if (typeof toRender === 'object' && !options) {
    options = toRender
    toRender = null
  }
  options = options || {}

  const getElement = () => (toRender ? toRender() : <Component />)

  it('implements system props', () => {
    for (const systemProps of systemPropArray) {
      expect(Component).toImplementSystemProps(systemProps)
    }
  })

  if (!options.skipSx) {
    it('implements the sx prop', () => {
      expect(Component).toImplementSxProp()
    })

    it('implements sx prop behavior', () => {
      expect(getElement()).toImplementSxBehavior()
    })
  }

  if (!options.skipAs) {
    it('respects the as prop', () => {
      const As = React.forwardRef((_props, ref) => <div className="as-component" ref={ref} />)
      const elem = React.cloneElement(getElement(), {as: As})
      expect(render(elem)).toEqual(render(<As />))
    })
  }

  it('sets a valid displayName', () => {
    expect(Component.displayName).toMatch(COMPONENT_DISPLAY_NAME_REGEX)
  })

  it('sets the default theme', () => {
    expect(getElement()).toSetDefaultTheme()
  })

  it('renders consistently', () => {
    expect(render(getElement())).toMatchSnapshot()
  })
}

export function checkExports(path, exports) {
  it('has declared exports', () => {
    const mod = require(`../${path}`)
    expect(mod).toSetExports(exports)
  })
}
