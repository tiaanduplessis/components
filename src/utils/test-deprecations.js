import semver from 'semver'
import {Deprecations} from '../utils/deprecate'

const ourVersion = require('../../package.json').version

beforeEach(() => {
  Deprecations.clearDeprecations()
})

afterEach(() => {
  const deprecations = Deprecations.getDeprecations()

  for (const dep of deprecations) {
    if (semver.gte(ourVersion, dep.version)) {
      throw new Error(`Found a deprecation that should be removed in ${dep.version}`)
    }
  }
})
