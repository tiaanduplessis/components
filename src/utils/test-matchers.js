import React from 'react'
import 'jest-styled-components'
import {styleSheetSerializer} from 'jest-styled-components/serializer'
import theme from '../theme'
import {getClasses, mount, getComputedStyles, render} from './testing'

expect.addSnapshotSerializer(styleSheetSerializer)

const stringify = d => JSON.stringify(d, null, '  ')

/**
 * These are props that styled-system aliases for backwards compatibility.
 * For some reason, they don't show up in our toImplementSystemProps() matcher,
 * so we skip over them.
 */
const ALIAS_PROP_TYPES = ['w', 'align', 'justify', 'wrap']

expect.extend({
  toMatchKeys(obj, values) {
    return {
      pass: Object.keys(values).every(key => this.equals(obj[key], values[key])),
      message: () => `Expected ${stringify(obj)} to have matching keys: ${stringify(values)}`
    }
  },

  toHaveClass(node, klass) {
    const classes = getClasses(node)
    const pass = classes.includes(klass)
    return {
      pass,
      message: () => `expected ${stringify(classes)} to include: ${stringify(klass)}`
    }
  },

  toHaveClasses(node, klasses, only = false) {
    const classes = getClasses(node)
    const pass = only ? this.equals(classes.sort(), klasses.sort()) : klasses.every(klass => classes.includes(klass))
    return {
      pass,
      message: () => `expected ${stringify(classes)} to include: ${stringify(klasses)}`
    }
  },

  toImplementSystemProps(Component, propNames) {
    const propKeys = new Set(Object.keys(Component.propTypes))
    const expectedPropKeys = Object.keys(propNames.propTypes)
    const missing = expectedPropKeys.filter(key => !propKeys.has(key)).filter(key => !ALIAS_PROP_TYPES.includes(key))
    return {
      pass: missing.length === 0,
      message: () => `Missing prop${missing.length === 1 ? '' : 's'}: ${stringify(missing)}`
    }
  },

  toImplementSxProp(Component) {
    return {
      pass: !!Component.propTypes.sx,
      message: () => 'Missing sx propTypes'
    }
  },

  toImplementSxBehavior(element) {
    const mediaKey = '@media (max-width:123px)'
    const sxPropValue = {
      [mediaKey]: {
        color: 'red.5'
      }
    }

    const elem = React.cloneElement(element, {sx: sxPropValue})
    const rendered = render(elem)

    function checkStylesDeep(rendered) {
      const className = rendered.props ? rendered.props.className : ''
      const styles = getComputedStyles(className)
      if (styles[mediaKey] && styles[mediaKey].color) {
        return true
      } else if (rendered.children) {
        return rendered.children.some(child => checkStylesDeep(child))
      } else {
        return false
      }
    }

    return {
      pass: checkStylesDeep(rendered),
      message: () => 'sx prop values did not change styles of component nor of any sub-components'
    }
  },

  toSetDefaultTheme(Component) {
    let comp
    if (Component.type) {
      comp = Component
    } else {
      comp = <Component />
    }
    const wrapper = mount(comp)
    const pass = this.equals(wrapper.prop('theme'), theme)
    return {
      pass,
      message: () => 'default theme is not set'
    }
  },

  toSetExports(mod, expectedExports) {
    if (!Object.keys(expectedExports).includes('default')) {
      return {
        pass: false,
        message: () => "You must specify the module's default export"
      }
    }

    const seen = new Set()
    for (const exp of Object.keys(expectedExports)) {
      seen.add(exp)
      if (mod[exp] !== expectedExports[exp]) {
        if (!mod[exp] && !expectedExports[exp]) {
          continue
        }

        return {
          pass: false,
          message: () => `Module exported a different value from key '${exp}' than expected`
        }
      }
    }

    for (const exp of Object.keys(mod)) {
      if (seen.has(exp)) {
        continue
      }

      if (mod[exp] !== expectedExports[exp]) {
        return {
          pass: false,
          message: () => `Module exported an unexpected value from key '${exp}'`
        }
      }
    }

    return {
      pass: true,
      message: () => ''
    }
  }
})
