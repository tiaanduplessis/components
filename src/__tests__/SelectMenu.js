import React from 'react'
import {SelectMenu, Button} from '..'
import {mount, render, renderRoot, COMPONENT_DISPLAY_NAME_REGEX, checkExports} from '../utils/testing'
import {COMMON} from '../constants'
import {render as HTMLRender, cleanup} from '@testing-library/react'
import {axe, toHaveNoViolations} from 'jest-axe'
import 'babel-polyfill'
expect.extend(toHaveNoViolations)

const BasicSelectMenu = ({onClick, as, align = 'left'}) => {
  return (
    <SelectMenu as={as}>
      <Button as="summary">Projects</Button>
      <SelectMenu.Modal title="Projects" align={align}>
        <SelectMenu.List>
          <SelectMenu.Item selected href="#">
            Primer Components bugs
          </SelectMenu.Item>
          <SelectMenu.Item onClick={onClick} as={as} data-test="menu-item" href="#">
            Primer Components roadmap
          </SelectMenu.Item>
          <SelectMenu.Divider>stuff</SelectMenu.Divider>
          <SelectMenu.Item href="#"> Project 3</SelectMenu.Item>
          <SelectMenu.Item href="#">Project 4</SelectMenu.Item>
          <SelectMenu.Footer>footer</SelectMenu.Footer>
        </SelectMenu.List>
      </SelectMenu.Modal>
    </SelectMenu>
  )
}

const MenuWithTabs = ({onClick}) => {
  return (
    <SelectMenu initialTab="Organization">
      <Button as="summary">Projects</Button>
      <SelectMenu.Modal title="Projects">
        <SelectMenu.Tabs>
          <SelectMenu.Tab index={0} onClick={onClick} data-test="repo-tab" tabName="Repository" />
          <SelectMenu.Tab index={1} tabName="Organization" />
        </SelectMenu.Tabs>
        <SelectMenu.TabPanel tabName="Repository">
          <SelectMenu.Item href="#">Primer Components bugs</SelectMenu.Item>
          <SelectMenu.Item href="#">Primer Components roadmap</SelectMenu.Item>
          <SelectMenu.Item href="#"> Project 3</SelectMenu.Item>
          <SelectMenu.Item href="#">Project 4</SelectMenu.Item>
        </SelectMenu.TabPanel>
        <SelectMenu.TabPanel tabName="Organization">
          <SelectMenu.Item href="#"> Project 2</SelectMenu.Item>
        </SelectMenu.TabPanel>
        <SelectMenu.Footer>Showing 3 of 3</SelectMenu.Footer>
      </SelectMenu.Modal>
    </SelectMenu>
  )
}

describe('SelectMenu', () => {
  checkExports('SelectMenu', {
    default: SelectMenu
  })

  for (const subComp of ['List', 'Divider', 'Filter', 'Item', 'List', 'Modal', 'Tabs', 'Tab', 'TabPanel', 'Header']) {
    const Comp = SelectMenu[subComp]

    it('implements the sx prop', () => {
      expect(Comp).toImplementSxProp()
    })

    it('sets a valid displayName', () => {
      expect(Comp.displayName).toMatch(COMPONENT_DISPLAY_NAME_REGEX)
    })
  }

  it('should have no axe violations', async () => {
    const {container} = HTMLRender(<BasicSelectMenu />)
    const results = await axe(container)
    expect(results).toHaveNoViolations()
    cleanup()
  })
  it('implements system props', () => {
    expect(SelectMenu).toImplementSystemProps(COMMON)
  })

  it('has default theme', () => {
    expect(SelectMenu).toSetDefaultTheme()
  })

  it('does not allow the "as" prop on SelectMenu', () => {
    const component = mount(<BasicSelectMenu as="span" />)
    expect(component.find('details').length).toEqual(1)
  })

  it('shows correct initial tab', () => {
    const testInstance = renderRoot(<MenuWithTabs />)
    expect(testInstance.findByProps({'aria-selected': true}).props.children).toBe('Organization')
  })

  it('clicking on a tab opens the tab', () => {
    const component = mount(<MenuWithTabs />)
    const tab = component.find("[data-test='repo-tab']").first()
    tab.simulate('click')
    expect(tab.getDOMNode().attributes['aria-selected']).toBeTruthy()
  })

  it('selected items have aria-checked', () => {
    const testInstance = renderRoot(<BasicSelectMenu />)
    expect(testInstance.findByProps({'aria-checked': true}).props.children[1]).toBe('Primer Components bugs')
  })

  it('clicking on a list item calls user provided onClick handler', () => {
    const mockClick = jest.fn()
    const component = mount(<BasicSelectMenu onClick={mockClick} />)
    const item = component.find("[data-test='menu-item']").first()
    item.simulate('click')
    expect(mockClick.mock.calls.length).toEqual(1)
  })

  it('clicking on a tab calls user provided onClick handler', () => {
    const mockClick = jest.fn()
    const component = mount(<MenuWithTabs onClick={mockClick} />)
    const item = component.find("[data-test='repo-tab']").first()
    item.simulate('click')
    expect(mockClick.mock.calls.length).toEqual(1)
  })

  it('clicking on an item closes the modal', () => {
    const component = mount(<BasicSelectMenu />)
    const item = component.find("[data-test='menu-item']").first()
    item.simulate('click')
    expect(component.getDOMNode().attributes.open).toBeFalsy()
  })

  it('right-aligned modal has right: 0px', () => {
    expect(render(<BasicSelectMenu align="right" />)).toMatchSnapshot()
  })
})
