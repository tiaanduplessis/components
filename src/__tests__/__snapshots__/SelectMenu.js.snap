// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectMenu right-aligned modal has right: 0px 1`] = `
.c1 {
  position: relative;
  display: inline-block;
  padding: 6px 16px;
  font-weight: 600;
  line-height: 20px;
  white-space: nowrap;
  vertical-align: middle;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border-radius: 6px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-align: center;
  font-size: 14px;
  color: #24292e;
  background-color: #fafbfc;
  border: 1px solid rgba(27,31,35,0.12);
  box-shadow: 0px 1px 0px rgba(27,31,35,0.04),inset 0px 2px 0px rgba(255,255,255,0.25);
}

.c1:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c1:focus {
  outline: none;
}

.c1:disabled {
  cursor: default;
}

.c1:disabled svg {
  opacity: 0.6;
}

.c1:hover {
  background-color: #F3F4F6;
  box-shadow: 0px 1px 0px rgba(209,213,218,0.2),inset 0px 2px 0px rgba(255,255,255,0.1);
}

.c1:focus {
  border-color: transparent;
  box-shadow: 0 0 0 3px rgba(3,102,214,0.3);
}

.c1:active {
  background-color: #edeff2;
  box-shadow: inset 0px 2px 0px rgba(149,157,165,0.1);
  border-color: #d1d5da;
}

.c1:disabled {
  color: #959da5;
  background-color: #fafbfc;
  border-color: #eaecef;
}

.c6 {
  padding: 4px 16px;
  margin: 0;
  font-size: 12px;
  font-weight: 600;
  color: #6a737d;
  background-color: #f6f8fa;
  border-bottom: 1px solid #eaecef;
}

.c7 {
  margin-top: -1px;
  padding: 8px 16px;
  font-size: 12px;
  color: #6a737d;
  text-align: center;
  border-top: 1px solid #e1e4e8;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 16px;
  overflow: hidden;
  text-align: left;
  cursor: pointer;
  background-color: #fff;
  border: 0;
  border-bottom: 1px solid #eaecef;
  color: #586069;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 12px;
  width: 100%;
}

.c5:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c5:focus {
  outline: none;
}

.c5[hidden] {
  display: none !important;
}

.c5 .SelectMenu-icon {
  width: 16px;
  margin-right: 8px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c5 .SelectMenu-selected-icon {
  visibility: hidden;
  -webkit-transition: -webkit-transform 0.12s cubic-bezier(0.5,0.1,1,0.5),visibility 0s 0.12s linear;
  -webkit-transition: transform 0.12s cubic-bezier(0.5,0.1,1,0.5),visibility 0s 0.12s linear;
  transition: transform 0.12s cubic-bezier(0.5,0.1,1,0.5),visibility 0s 0.12s linear;
  -webkit-transform: scale(0);
  -ms-transform: scale(0);
  transform: scale(0);
}

.c5[aria-checked='true'] {
  font-weight: 500;
  color: #24292e;
}

.c5[aria-checked='true'] .SelectMenu-selected-icon {
  visibility: visible;
  -webkit-transition: -webkit-transform 0.12s cubic-bezier(0,0,0.2,1),visibility 0s linear;
  -webkit-transition: transform 0.12s cubic-bezier(0,0,0.2,1),visibility 0s linear;
  transition: transform 0.12s cubic-bezier(0,0,0.2,1),visibility 0s linear;
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.c4 {
  position: relative;
  padding: 0;
  margin: 0;
  -webkit-flex: auto;
  -ms-flex: auto;
  flex: auto;
  overflow-x: hidden;
  overflow-y: auto;
  background-color: #fff;
  -webkit-overflow-scrolling: touch;
}

.c3 {
  position: relative;
  z-index: 99;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  max-height: 66%;
  margin: auto 0;
  overflow: hidden;
  pointer-events: auto;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  background-color: #fff;
  border-radius: 6px;
  box-shadow: 0 1px 5px rgba(27,31,35,0.15);
  -webkit-animation: lejQAW 0.12s cubic-bezier(0,0.1,0.1,1) backwards;
  animation: lejQAW 0.12s cubic-bezier(0,0.1,0.1,1) backwards;
}

.c2 {
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: 99;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 16px;
  pointer-events: none;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c2::before {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  pointer-events: none;
  content: '';
  background-color: rgba(27,31,35,0.5);
}

.c0[open] > summary::before {
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: 80;
  display: block;
  cursor: default;
  content: ' ';
  background: transparent;
}

.c0 > summary {
  list-style: none;
}

.c0 > summary::before {
  display: none;
}

.c0 > summary::-webkit-details-marker {
  display: none;
}

@media (min-width:544px) {
  .c7 {
    padding: 4px 8px;
  }
}

@media (min-width:544px) {
  .c5 {
    padding-top: 8px;
    padding-bottom: 8px;
  }
}

@media (hover:hover) {
  .c5 body:not(.intent-mouse) .SelectMenu-item:focus,
  .c5:hover,
  .c5:active,
  .c5:focus {
    background-color: #f6f8fa;
  }
}

@media (hover:none) {
  .c5 {
    -webkit-tap-highlight-color: rgba(#d1d5da,0.5);
  }

  .c5:focus,
  .c5:active {
    background-color: #fafbfc;
  }
}

@media (hover:hover) {
  .c4 .SelectMenuTab:focus {
    background-color: #dbedff;
  }

  .c4 .SelectMenuTab:not([aria-checked='true']):hover {
    color: #24292e;
    background-color: #e1e4e8;
  }

  .c4 .SelectMenuTab:not([aria-checked='true']):active {
    color: #24292e;
    background-color: #f6f8fa;
  }
}

@media (min-width:544px) {
  .c3 {
    width: '300px';
    height: auto;
    max-height: 350px;
    margin: 4px 0 16px 0;
    font-size: 12px;
    border: 1px solid #d1d5da;
    border-radius: 6px;
    box-shadow: 0 1px 5px rgba(27,31,35,0.15) !default;
  }
}

@media (min-width:544px) {
  .c2::before {
    display: none;
  }
}

@media (min-width:544px) {
  .c2 {
    position: absolute;
    top: auto;
    right: 0;
    bottom: auto;
    left: auto;
    padding: 0;
  }
}

<details
  className="c0"
  onToggle={[Function]}
  open={false}
>
  <summary
    className="c1"
  >
    Projects
  </summary>
  <div
    className="c2"
    role="menu"
    title="Projects"
  >
    <div
      className="c3"
    >
      <div
        className="c4"
      >
        <a
          aria-checked={true}
          className="c5"
          href="#"
          onClick={[Function]}
          role="menuitemcheckbox"
        >
          <svg
            aria-hidden="true"
            className="SelectMenu-icon SelectMenu-selected-icon"
            dangerouslySetInnerHTML={
              Object {
                "__html": "<path fill-rule=\\"evenodd\\" d=\\"M13.78 4.22a.75.75 0 010 1.06l-7.25 7.25a.75.75 0 01-1.06 0L2.22 9.28a.75.75 0 011.06-1.06L6 10.94l6.72-6.72a.75.75 0 011.06 0z\\"></path>",
              }
            }
            fill="currentColor"
            height={16}
            role="img"
            style={
              Object {
                "display": "inline-block",
                "userSelect": "none",
                "verticalAlign": "text-bottom",
              }
            }
            viewBox="0 0 16 16"
            width={16}
          />
          Primer Components bugs
        </a>
        <a
          aria-checked={false}
          className="c5"
          data-test="menu-item"
          href="#"
          onClick={[Function]}
          role="menuitemcheckbox"
        >
          <svg
            aria-hidden="true"
            className="SelectMenu-icon SelectMenu-selected-icon"
            dangerouslySetInnerHTML={
              Object {
                "__html": "<path fill-rule=\\"evenodd\\" d=\\"M13.78 4.22a.75.75 0 010 1.06l-7.25 7.25a.75.75 0 01-1.06 0L2.22 9.28a.75.75 0 011.06-1.06L6 10.94l6.72-6.72a.75.75 0 011.06 0z\\"></path>",
              }
            }
            fill="currentColor"
            height={16}
            role="img"
            style={
              Object {
                "display": "inline-block",
                "userSelect": "none",
                "verticalAlign": "text-bottom",
              }
            }
            viewBox="0 0 16 16"
            width={16}
          />
          Primer Components roadmap
        </a>
        <div
          className="c6"
        >
          stuff
        </div>
        <a
          aria-checked={false}
          className="c5"
          href="#"
          onClick={[Function]}
          role="menuitemcheckbox"
        >
          <svg
            aria-hidden="true"
            className="SelectMenu-icon SelectMenu-selected-icon"
            dangerouslySetInnerHTML={
              Object {
                "__html": "<path fill-rule=\\"evenodd\\" d=\\"M13.78 4.22a.75.75 0 010 1.06l-7.25 7.25a.75.75 0 01-1.06 0L2.22 9.28a.75.75 0 011.06-1.06L6 10.94l6.72-6.72a.75.75 0 011.06 0z\\"></path>",
              }
            }
            fill="currentColor"
            height={16}
            role="img"
            style={
              Object {
                "display": "inline-block",
                "userSelect": "none",
                "verticalAlign": "text-bottom",
              }
            }
            viewBox="0 0 16 16"
            width={16}
          />
           Project 3
        </a>
        <a
          aria-checked={false}
          className="c5"
          href="#"
          onClick={[Function]}
          role="menuitemcheckbox"
        >
          <svg
            aria-hidden="true"
            className="SelectMenu-icon SelectMenu-selected-icon"
            dangerouslySetInnerHTML={
              Object {
                "__html": "<path fill-rule=\\"evenodd\\" d=\\"M13.78 4.22a.75.75 0 010 1.06l-7.25 7.25a.75.75 0 01-1.06 0L2.22 9.28a.75.75 0 011.06-1.06L6 10.94l6.72-6.72a.75.75 0 011.06 0z\\"></path>",
              }
            }
            fill="currentColor"
            height={16}
            role="img"
            style={
              Object {
                "display": "inline-block",
                "userSelect": "none",
                "verticalAlign": "text-bottom",
              }
            }
            viewBox="0 0 16 16"
            width={16}
          />
          Project 4
        </a>
        <footer
          className="c7"
        >
          footer
        </footer>
      </div>
    </div>
  </div>
</details>
`;
